### PR TITLE
FPv2: add timing unit test

### PR DESCRIPTION
**Pull request by sshane**
Thought this would be nice to ground truth any investigation into slow FW query complaints. I had no idea what the expected time to fingerprint for all the brands were before writing this. See output below.

### CODE REVIEW

**Comment by sshane** [comment]
```python
batman@z840-openpilot3:~/openpilot/selfdrive/car/tests$ ./test_fw_fingerprint_metrics.py 
brand='body', num_pandas=1, len(requests)=1, total FW query time=0.1 seconds
brand='chrysler', num_pandas=1, len(requests)=3, total FW query time=0.3 seconds
brand='ford', num_pandas=1, len(requests)=2, total FW query time=0.2 seconds
brand='honda', num_pandas=1, len(requests)=4, total FW query time=0.5 seconds
brand='hyundai', num_pandas=1, len(requests)=6, total FW query time=0.7 seconds
brand='mazda', num_pandas=1, len(requests)=1, total FW query time=0.1 seconds
brand='nissan', num_pandas=1, len(requests)=3, total FW query time=0.3 seconds
brand='subaru', num_pandas=1, len(requests)=1, total FW query time=0.1 seconds
brand='tesla', num_pandas=1, len(requests)=2, total FW query time=0.2 seconds
brand='toyota', num_pandas=1, len(requests)=3, total FW query time=0.3 seconds
brand='volkswagen', num_pandas=1, len(requests)=2, total FW query time=0.2 seconds
.brand='hyundai', num_pandas=2, len(requests)=10, total FW query time=1.1 seconds
```

**Comment by adeebshihadeh** [comment]
Nice, can we also get a worse  case for CAN fingerprinting cars?

**Comment by sshane** [comment]
This actually doesn't account for things like the Toyota ECUs with subaddresses needing to be queried individually. We can get more accurate times by just calling the functions themselves with fake sockets, or real sockets with empty data and letting it timeout. I'll adapt test_startup for this.

**Comment by adeebshihadeh** [comment]
Can you finish this up?

**Comment by sshane** [comment]
With thread and fake socket:

```python
brand='body', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=1, total FW query time=0.11 seconds
brand='chrysler', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=3, total FW query time=0.3 seconds
brand='ford', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=2, total FW query time=0.2 seconds
brand='honda', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=4, total FW query time=0.48 seconds
brand='hyundai', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=10, total FW query time=0.75 seconds
brand='mazda', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=1, total FW query time=0.18 seconds
brand='nissan', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=3, total FW query time=0.3 seconds
brand='subaru', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=1, total FW query time=0.1 seconds
brand='tesla', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=2, total FW query time=0.2 seconds
brand='toyota', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=3, total FW query time=0.71 seconds
brand='volkswagen', num_pandas=1, len(FW_QUERY_CONFIGS[brand].requests)=2, total FW query time=0.2 seconds
brand='hyundai', num_pandas=2, len(FW_QUERY_CONFIGS[brand].requests)=10, total FW query time=1.13 seconds
```

File: selfdrive/car/tests/test_fw_fingerprint.py [hunk 1]
```diff
@@ -16,6 +19,14 @@
 VERSIONS = get_interface_attr("FW_VERSIONS", ignore_none=True)
 
 
+class FakeSocket:
```
Comment (sshane) [inline selfdrive/car/tests/test_fw_fingerprint.py:22]: The one in process replay either hangs or throws an exception with the empty list being popped from. We really just want these functions to do nothing so the timeouts in IsoTpParallelQuery are hit

File: selfdrive/car/tests/test_fw_fingerprint.py [hunk 2]
```diff
@@ -104,5 +115,27 @@ def test_fw_request_ecu_whitelist(self):
...
+        thread.start()
+
+        t = time.perf_counter()
+        while thread.is_alive():
+          time.sleep(0.01)
+          if not params.get_bool("ObdMultiplexingChanged"):
+            params.put_bool("ObdMultiplexingChanged", True)
+
+        total_time = round(time.perf_counter() - t, 2)
+        self.assertLessEqual(total_time, 1.3)
```
Comment (adeebshihadeh) [inline selfdrive/car/tests/test_fw_fingerprint.py:136]: let's add a min time too, so that we have to update the test thresholds if it improves

Comment (adeebshihadeh) [reply selfdrive/car/tests/test_fw_fingerprint.py:136]: the thresholds should also be brand specific

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]

File: selfdrive/car/tests/test_fw_fingerprint.py (hunk 2 above)
Comment (sshane) [reply selfdrive/car/tests/test_fw_fingerprint.py]: done, added the theoretical calculated time with some tolerance

File: selfdrive/car/tests/test_fw_fingerprint.py [hunk 3]
```diff
@@ -104,5 +115,55 @@ def test_fw_request_ecu_whitelist(self):
...
+      thread = threading.Thread(target=get_fw_versions, args=(fake_socket, fake_socket, brand), kwargs=dict(num_pandas=num_pandas))
+      thread.start()
+      while thread.is_alive():
+        time.sleep(0.01)
+        if not params.get_bool("ObdMultiplexingChanged"):
+          params.put_bool("ObdMultiplexingChanged", True)
+
+    avg_time = round((time.perf_counter() - t) / n, 2)
+    self.assertLess(avg_time, ref_time + tol)
+    self.assertGreater(avg_time, ref_time - tol, "Performance seems to have improved, update test references.")
```
Comment (sshane) [inline selfdrive/car/tests/test_fw_fingerprint.py:134]:
```suggestion
    self.assertGreater(avg_time, ref_time - tol, "Performance seems to have improved, update test refs.")
```

**Pull request merged**